flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/parser/SqlClientSyntaxHighlighter.java

```diff
@@ -0,0 +1,246 @@
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.table.client.cli.CliClient.COLOR_SCHEMA_VAR;
+
+/** Sql Client syntax highlighter. */
+public class SqlClientSyntaxHighlighter extends DefaultHighlighter {
+    private static final Set<String> FLINK_KEYWORD_SET;
+    private static final Set<String> HIVE_KEYWORD_SET;
+
+    static {
+        try (InputStream is =
+                SqlClientSyntaxHighlighter.class.getResourceAsStream("/keywords.properties")) {
+            Properties props = new Properties();
+            props.load(is);
+            FLINK_KEYWORD_SET =
+                    Collections.unmodifiableSet(
+                            Arrays.stream(props.get("default").toString().split(";"))
+                                    .collect(Collectors.toSet()));
+            HIVE_KEYWORD_SET =
+                    Collections.unmodifiableSet(
+                            Arrays.stream(props.get("hive").toString().split(";"))
+                                    .collect(Collectors.toSet()));
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private final Executor executor;
+
+    public SqlClientSyntaxHighlighter(Executor executor) {
+        this.executor = executor;
+    }
+
+    @Override
+    public AttributedString highlight(LineReader reader, String buffer) {
+
+        final Object colorSchemeOrdinal = reader.getVariable(COLOR_SCHEMA_VAR);
+        SyntaxHighlightStyle.BuiltInStyle style =
+                SyntaxHighlightStyle.BuiltInStyle.fromOrdinal(
+                        colorSchemeOrdinal == null ? 0 : (Integer) colorSchemeOrdinal);
+        if (style == SyntaxHighlightStyle.BuiltInStyle.DEFAULT) {
+            return super.highlight(reader, buffer);
+        }
+        final SqlDialect dialect =
+                executor.getSessionConfig()
+                                .get(TableConfigOptions.TABLE_SQL_DIALECT)
+                                .equalsIgnoreCase(SqlDialect.HIVE.toString())
+                        ? SqlDialect.HIVE
+                        : SqlDialect.DEFAULT;
+        return getHighlightedOutput(buffer, style.getHighlightStyle(), dialect);
+    }
+
+    static AttributedString getHighlightedOutput(
+            String buffer, SyntaxHighlightStyle style, SqlDialect dialect) {
+        final Set<Character> stateStartSymbols =
+                Arrays.stream(State.values())
+                        .map(t -> t.getStart(dialect).charAt(0))
+                        .collect(Collectors.toSet());
+        final AttributedStringBuilder highlightedOutput = new AttributedStringBuilder();
+        State currentParseState = null;
+        int counter = 0;
+        StringBuilder word = new StringBuilder();
+        for (int i = 0; i < buffer.length(); i++) {
+            final char c = buffer.charAt(i);
+            if (currentParseState == null) {
+                if (stateStartSymbols.contains(c)) {
```

Comment:
I know escaping with doubling of symbols e.g. `''`, `""` - it works in Flink.
Also some engines support escaping with help of backslash `\` however it seems Flink does not support it (yet)
e.g. query
```sql
SELECT '\'; --returns \ as a varchar
```
```sql
SELECT 123 AS `\`; --returns 123 with a field named as \
```
```sql
SELECT ''''; -- returns value with 1 quote
```
```sql
SELECT 1 AS ````; -- returns a value with field name containing a tick
```
in case there is no support for escaping with backslash it should be ok, since 

```sql
-- for the case of escaping quotes or ticks (depending on the dialect)
select ''''; -- all 4 single quotes should be in quoted style
select 'test '' 123 '; -- everything between first and last singe quote should be in a quoted style
```

Or did I miss anything?

Reply:
Thanks for the sharing. Can we add these in the tests?

Reply:
yes, sure
I added more tests including mentioned here, also tried to make adding tests for syntax highlighting simpler